pkg/apis/events/v1/cloudstoragesource_validation.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2019 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/google/knative-gcp/pkg/apis/duck"
+)
+
+func (current *CloudStorageSource) Validate(ctx context.Context) *apis.FieldError {
+	errs := current.Spec.Validate(ctx).ViaField("spec")
+	return duck.ValidateAutoscalingAnnotations(ctx, current.Annotations, errs)
+}
+
+func (current *CloudStorageSourceSpec) Validate(ctx context.Context) *apis.FieldError {
+	var errs *apis.FieldError
+
+	// Sink [required]
+	if equality.Semantic.DeepEqual(current.Sink, duckv1.Destination{}) {
+		errs = errs.Also(apis.ErrMissingField("sink"))
+	} else if err := current.Sink.Validate(ctx); err != nil {
+		errs = errs.Also(err.ViaField("sink"))
+	}
+
+	// Bucket [required]
+	if current.Bucket == "" {
+		errs = errs.Also(apis.ErrMissingField("bucket"))
+	}
+
+	if err := duck.ValidateCredential(current.Secret, current.ServiceAccountName); err != nil {
+		errs = errs.Also(err)
+	}
+
+	return errs
+}
+
+func (current *CloudStorageSource) CheckImmutableFields(ctx context.Context, original *CloudStorageSource) *apis.FieldError {
+	if original == nil {
+		return nil
+	}
+	// Modification of EventType, Secret, ServiceAccount, Project, Bucket, ObjectNamePrefix and PayloadFormat are not allowed. Everything else is mutable.
```

Comment:
nit: remove PayloadFormat from this list

Reply:
serviceAccount is not there anymore I think...

Reply:
There was also an ugly bug in eventing where CheckImmutableFields was not being called for some reason. 
Can you check with v1beta1 if it is actually working?

Reply:
better yet, once v1 is ready, check that this callback is actually being called

Reply:
Just double checked with @grac3gao, the `serviceAccountName` should be immutable. I made a fix for all API versions in the new commit.